File: app/src/main/java/com/example/chgadapter_kotlin_demo/weibo/VH/SourceViewHolder.java
//package com.example.chgadapter_kotlin_demo.weibo.VH;
//
//
//import android.content.Context;
//import android.graphics.drawable.Drawable;
//import android.util.Log;
//import android.view.View;
//import android.view.ViewGroup;
//import android.widget.ImageView;
//import android.widget.LinearLayout;
//
//import androidx.annotation.NonNull;
//
//import com.bumptech.glide.Glide;
//import com.bumptech.glide.load.engine.DiskCacheStrategy;
//import com.bumptech.glide.request.RequestOptions;
//import com.bumptech.glide.request.target.SimpleTarget;
//import com.bumptech.glide.request.transition.Transition;
//import com.chg.adapter.EventTransmissionListener;
//import com.chg.adapter.Model;
//import com.chg.adapter.ViewHolder;
//import com.example.chgadapter_kotlin_demo.R;
//import com.example.chgadapter_kotlin_demo.weibo.model.FoundSendData;
//import com.example.chgadapter_kotlin_demo.weibo.model.Source;
//
//import java.util.HashMap;
//
//public class SourceViewHolder extends ViewHolder<Source> {
//
//    private ImageView imageView;
//
//    public SourceViewHolder(@NonNull View itemView, EventTransmissionListener eventTransmissionListener, ViewGroup parent) {
//        super(itemView, eventTransmissionListener, parent);
//        imageView = findViewById(R.id.imageView);
//
//        itemView.setOnClickListener(new View.OnClickListener() {
//            @Override
//            public void onClick(View v) {
//                if (getCustomData() != null) {
//                    final FoundSendData foundSendData = (FoundSendData) getCustomData();
//                    HashMap map = new HashMap();
//                    map.put("position", getAdapterPosition());
//                    map.put("sources", foundSendData.getContent().getSource());
//                    getEventTransmissionListener().onEventTransmission(this, map, 0, null);
//                } else {
//                    getEventTransmissionListener().onEventTransmission(this, getModel(), 0, null);
//                }
//            }
//        });
//    }
//
//    @Override
//    public void onCreated() {
//
//    }
//
//    @Override
//    public void onBindViewHolder(final Source model) {
//        super.onBindViewHolder(model);
//        String url = getUrl(getModel(), getPicWidth(getModel()));
//        model.setHandleUrl(url);
//    }
//
//    @Override
//    public void onViewAttachedToWindow() {
//        super.onViewAttachedToWindow();
//        Source source = (Source) getModel();
//        String url = source.getHandleUrl();
//        imageView.setTag(url);
//        displayImageCente2(imageView, url, getContext(), R.drawable.default_pic, false);
//    }
//
//    public void displayImageCente2(final ImageView imageview, final String url, Context context, int defultPic, Boolean isCircleCrop) {
//        RequestOptions options = null;
//        if (isCircleCrop) {
//            options = new RequestOptions().diskCacheStrategy(DiskCacheStrategy.RESOURCE).placeholder(defultPic).error(defultPic).dontAnimate().circleCropTransform();
//        } else {
//            options = new RequestOptions().diskCacheStrategy(DiskCacheStrategy.RESOURCE).placeholder(defultPic).error(defultPic).dontAnimate();
//        }
//
//        Glide.with(context).load(url).apply(options).into(new SimpleTarget<Drawable>() {
//            @Override
//            public void onResourceReady(Drawable drawable, Transition<? super Drawable> transition) {
//                if (imageview != null) {
//                    String urlT = (String) imageview.getTag();
//                    if (drawable != null && urlT != null && urlT.equals(url)) {
//                        imageview.setImageDrawable(drawable);
//                    }
//                }
//            }
//        });
//    }
//
//    @Override
//    public void onViewDetachedFromWindow() {
//        super.onViewDetachedFromWindow();
//        Glide.with(getContext()).clear(imageView);
//        imageView.setImageResource(R.drawable.default_pic);
//    }
//
//    //获取图片url
//    public String getUrl(Model model, int imageWidth) {
//        String url = null;
//        if (getCustomData() != null) {
//            final FoundSendData foundSendData = (FoundSendData) getCustomData();
//            Integer type = foundSendData.getContent().getType();
//            if (type == 2) {//图片
//                Source source = (Source) model;
//                url = source.getUrl() + "?x-oss-process=image/resize,w_" + imageWidth + "/quality,q_50";
//            } else if (type == 3) {//视频
//                url = foundSendData.getContent().getCover();
//            }
//        } else {
//            Source source = (Source) model;
//            int sourceType = source.getSourceType();
//            if (sourceType == 4) {//视频
//                url = source.getUrl();
//            } else {
//                url = source.getUrl() + "?x-oss-process=image/resize,w_" + imageWidth + "/quality,q_50";
//            }
//        }
//
//        return url;
//    }
//
//    //获取图片的宽高
//    public int getPicWidth(Model model) {
//        int imageWidth = 0;
//        int viewWidth = getContext().getResources().getDisplayMetrics().widthPixels;
//        int viewHeight = getContext().getResources().getDisplayMetrics().heightPixels;
//        Source source = (Source) model;
//        Log.i("chg", "source.getHeight():" + source.getHeight() + "   source.getWidth()" + source.getWidth() + "    url:" + source.getUrl());
//        if (getCustomData() != null) {
//            final FoundSendData foundSendData = (FoundSendData) getCustomData();
//            if (foundSendData.getContent().getSource() != null && foundSendData.getContent().getSource().size() > 0) {
//                int size = foundSendData.getContent().getSource().size();
//                if (size == 1) {
//                    if (source.getHeight() == 0 || source.getWidth() == 0) {
//                        imageView.setLayoutParams(new LinearLayout.LayoutParams(viewWidth, viewWidth));
//                        imageWidth = viewWidth;
//                    } else {
//                        int height = (int) (source.getHeight() / source.getWidth() * viewWidth);
//                        height = height >= viewHeight ? (int) (viewHeight * 0.7) : height;
//                        imageView.setLayoutParams(new LinearLayout.LayoutParams(viewWidth, height));
//                        imageWidth = viewWidth;
//                    }
//                } else if (size == 2) {
//                    imageView.setLayoutParams(new LinearLayout.LayoutParams(viewWidth / 2, viewWidth / 2));
//                    imageWidth = viewWidth / 2;
//                } else {
//                    imageView.setLayoutParams(new LinearLayout.LayoutParams(viewWidth / 3, viewWidth / 3));
//                    imageWidth = viewWidth / 3;
//                }
//            }
//        } else {
//            imageView.setLayoutParams(new LinearLayout.LayoutParams(viewWidth, viewHeight));
//            imageWidth = viewWidth;
//        }
//
//        imageWidth = imageWidth > 900 ? 900 : imageWidth;
//        return (int) imageWidth;
//    }
//}

File: app/src/main/java/com/example/chgadapter_kotlin_demo/weibo/VH/SourceViewHolder.kt
package com.example.chgadapter_kotlin_demo.weibo.VH

import android.content.Context
import android.graphics.drawable.Drawable
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.request.RequestOptions
import com.bumptech.glide.request.target.SimpleTarget
import com.bumptech.glide.request.transition.Transition
import com.chg.adapter.base.EventTransmissionListener
import com.chg.adapter.base.Model
import com.chg.adapter.recyclerView.ViewHolder
import com.example.chgadapter_kotlin_demo.R
import com.example.chgadapter_kotlin_demo.weibo.model.FoundSendData
import com.example.chgadapter_kotlin_demo.weibo.model.Source
import java.util.*

class SourceViewHolder(
    itemView: View,
    eventTransmissionListener: EventTransmissionListener?,
    parent: ViewGroup?
) : ViewHolder<Source>(itemView, eventTransmissionListener, parent) {

    private lateinit var imageView: ImageView


    override fun onCreated() {
        imageView = findViewById<ImageView>(R.id.imageView)
        itemView.setOnClickListener(object : View.OnClickListener {
            override fun onClick(v: View) {
                Log.i("chgLog","点击图片")
                if (getCustomData() != null) {
                    val foundSendData = getCustomData() as FoundSendData?
                    var map = mutableMapOf<Any,Any>()
                    map["position"] = adapterPosition
                    map["sources"] = foundSendData!!.content.source
                    eventTransmission(this@SourceViewHolder,map,0,null)
                } else {
                    eventTransmission(this@SourceViewHolder,getModel(),0,null)
                }
            }
        })
    }

    override fun onBindViewHolder(model: Source?) {
        super.onBindViewHolder(model)
        val url = getUrl(getModel(), getPicWidth(getModel()))
        if (model != null) {
            model.handleUrl = url
        }
    }

    override fun onViewAttachedToWindow() {
        super.onViewAttachedToWindow()
        val url = getModel()?.handleUrl
        imageView!!.tag = url
        displayImageCente2(imageView, url, getContext(), R.drawable.default_pic, false)
    }

    fun displayImageCente2(
        imageview: ImageView?,
        url: String?,
        context: Context?,
        defultPic: Int,
        isCircleCrop: Boolean
    ) {
        var options: RequestOptions? = null
        options = if (isCircleCrop) {
            RequestOptions.circleCropTransform()
        } else {
            RequestOptions().diskCacheStrategy(DiskCacheStrategy.RESOURCE).placeholder(defultPic)
                .error(defultPic).dontAnimate()
        }
        Glide.with(context!!).load(url).apply(options).into(object : SimpleTarget<Drawable?>() {
            override fun onResourceReady(
                resource: Drawable,
                transition: Transition<in Drawable?>?
            ) {
                if (imageview != null) {
                    if (imageview.tag != null) {
                        val urlT = imageview.tag as String
                        if (resource != null && urlT != null && urlT == url) {
                            imageview.setImageDrawable(resource)
                        }
                    }
                }
            }
        })
    }

    override fun onViewDetachedFromWindow() {
        super.onViewDetachedFromWindow()
        Glide.with(getContext()).clear(imageView!!)
        imageView!!.setImageResource(R.drawable.default_pic)
    }

    //获取图片url
    fun getUrl(model: Model?, imageWidth: Int): String? {
        var url: String? = null
        if (getCustomData() != null) {
            val foundSendData = getCustomData() as FoundSendData?
            val type = foundSendData!!.content.type
            if (type == 2) { //图片
                val source = model as Source?
                url = source!!.url + "?x-oss-process=image/resize,w_" + imageWidth + "/quality,q_50"
            } else if (type == 3) { //视频
                url = foundSendData.content.cover
            }
        } else {
            val source = model as Source?
            val sourceType = source!!.sourceType
            url = if (sourceType == 4) { //视频
                source.url
            } else {
                source.url + "?x-oss-process=image/resize,w_" + imageWidth + "/quality,q_50"
            }
        }
        Log.i("chgLog","url:$url")
        return url
    }

    //获取图片的宽高
    fun getPicWidth(model: Model?): Int {
        var imageWidth = 0
        val viewWidth = getContext().resources.displayMetrics.widthPixels
        val viewHeight = getContext().resources.displayMetrics.heightPixels
        val source = model as Source?
        Log.i(
            "chg",
            "source.getHeight():" + source!!.height + "   source.getWidth()" + source.width + "    url:" + source.url
        )
        if (getCustomData() != null) {
            val foundSendData = getCustomData() as FoundSendData?
            if (foundSendData!!.content.source != null && foundSendData.content.source.size > 0) {
                val size = foundSendData.content.source.size
                if (size == 1) {
                    if (source.height == 0f || source.width == 0f) {
                        imageView!!.layoutParams = LinearLayout.LayoutParams(viewWidth, viewWidth)
                        imageWidth = viewWidth
                    } else {
                        var height = (source.height / source.width * viewWidth).toInt()
                        height = if (height >= viewHeight) (viewHeight * 0.7).toInt() else height
                        imageView!!.layoutParams = LinearLayout.LayoutParams(viewWidth, height)
                        imageWidth = viewWidth
                    }
                } else if (size == 2) {
                    imageView!!.layoutParams =
                        LinearLayout.LayoutParams(viewWidth / 2, viewWidth / 2)
                    imageWidth = viewWidth / 2
                } else {
                    imageView!!.layoutParams =
                        LinearLayout.LayoutParams(viewWidth / 3, viewWidth / 3)
                    imageWidth = viewWidth / 3
                }
            }
        } else {
            imageView!!.layoutParams = LinearLayout.LayoutParams(viewWidth, viewHeight)
            imageWidth = viewWidth
        }
        imageWidth = if (imageWidth > 900) 900 else imageWidth
        return imageWidth
    }
}